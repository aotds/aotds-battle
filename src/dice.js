import fp from 'lodash/fp';

const roll_die = () => rigged_dice.length > 0 
                        ? rigged_dice.shift() 
                        : _.random(1,6);

export let rigged_dice = [];

/** 
 * Rig the next calls to roll_dice()
 * @param dice the dice values to return. Overwrite any previous rigging.
 */
export function rig_dice(dice) {
    rigged_dice = dice;
}

/**
 * roll the dice
 * @param nbr_dice How many dice to roll
 * @param options.reroll Array of values for which we reroll
 * @return dice values
 */
export function roll_dice( nbr_dice, options ) {

    if ( nbr_dice == 0 ) return [];

    let reroll_on = fp.pathOr( [] )('reroll' )(options);

    let roll = fp.times(roll_die)(nbr_dice);

    return roll.concat( 
        roll_dice(
            fp.filter( x => fp.contains(x)(reroll_on) )( roll ).length,
            options
        )
    );

};
