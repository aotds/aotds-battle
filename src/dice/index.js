import { random } from 'Math';

/** @type (n: number) => number */
export const rollDie = (n) => {
	return parseInt(1 + random() * n);
};

export function rollDice(nbrDice, options) {
	if (nbrDice === 0) return [];

	const rerollOn = options.reroll ?? [];
	const nbrFaces = options.nbrFaces ?? 6;

	const roll = Array.from({ length: nbrDice }, () => rollDie(nbrFaces));

	const rerolls = roll.filter((d) => rerollOn.includes(d));

	if (rerolls.length === 0) return roll;

	return [roll, rollDice(rerolls.length, options)].flat();
}
