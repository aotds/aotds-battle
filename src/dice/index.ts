import  random  from 'lodash/random.js';

const rollDie = (n: number) => random(1, n);

export function rollDice(
	nbr_dice: number,
	options: { reroll?: number[]; nbr_faces?: number; note?: unknown } = {},
): number[] {
	if (nbr_dice === 0) return [];

	const reroll_on = options.reroll ?? [];
	const nbr_faces = options.nbr_faces ?? 6;

	const roll: number[] = Array.from({ length: nbr_dice }, () =>
		rollDie(nbr_faces),
	);

	const rerolls = roll.filter((d) => (reroll_on as any).includes(d));

	if (rerolls.length === 0) return roll;

	return ([roll, rollDice(rerolls.length, options)] as any).flat();
}
