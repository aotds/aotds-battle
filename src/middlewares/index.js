import { add_timestamp, add_action_id, add_parent_action } from './meta';
import validate_schema from './validate_schema';
import play_turn from './play_turn';

import weapons_mw from './weapons';
import movement_mw from './movement';

// players
//Check all ships
//ships with orders not done
//filter those not associated with players
//filter those not associated with active players
//make sure there are at least 2 active players 
//if 
const debug = require('debug')('aotds:mw');


const trycatch = ({getState}) => next => action => {
    let state = getState();
    try {
        next(action);
    }
    catch(e) {
        debug(action);
        debug(state);
        throw e;
    }
};

export default [
    add_timestamp,
    add_action_id,
    add_parent_action,
    validate_schema,
    play_turn,
    ...weapons_mw,
    ...movement_mw,
    trycatch,
];
