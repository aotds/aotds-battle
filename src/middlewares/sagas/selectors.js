import _ from 'lodash';

export const object_by_id = ( store, id ) => _.find( store.objects, { id } );
