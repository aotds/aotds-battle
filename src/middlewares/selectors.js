import _ from 'lodash';
import fp from 'lodash/fp';

const debug = require('debug')('aotds:mw:selector');

export function select( func, ...args ) {
    return func(...args)
}

export const get_bogeys = () => _.flow([
    fp.getOr({})('bogeys'),
    fp.values,
]);

export const get_bogey = id => fp.flow([
    fp.getOr([],'bogeys'),
    fp.find({ id })
]);

export function get_active_players(state) {
    return fp.reject( { status: 'inactive' } )( state.game.players );
};

export function get_players_not_done( state ) {
    let active_players = get_active_players(state);

    // all bogeys not done 
    let players_not_done = state 
        |> fp.get('bogeys')
        |> fp.reject('orders.done')
        |> fp.map('player_id')
        |> fp.uniq;

    return active_players |> fp.filter( p => _.includes(players_not_done,p.id) );
}

