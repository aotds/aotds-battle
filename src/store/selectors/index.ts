import _ from 'lodash';
import fp from 'lodash/fp';

import { BattleState } from "../types";

export function get_active_players(state: BattleState): string[] {
    return _(state.game.players).reject( 'inactive' ).map( 'id' ).value();
};

export function get_players_not_done( state: BattleState ): string[] {
    let active_players = get_active_players(state);

    // all bogeys not done
    let players_not_done = fp.flow([
        fp.get('bogeys'),
        fp.reject('orders.done'),
        fp.map('player_id'),
        fp.uniq
    ])(state)

    return _.intersection( players_not_done, active_players );
}
