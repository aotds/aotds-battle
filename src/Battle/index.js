// @flow

import { createStore, applyMiddleware } from "redux";
import createSagaMiddleware from 'redux-saga';

import _ from 'lodash';

import logger from '../Logger';
import reducer from './Reducer';

import type { BattleState, BattleAction } from './Types';

import actions from './Actions';
import sagas from './Sagas';

type InitOptions = {
    name: string
};

const MW_logger = store => next => action => {
    logger.trace({ action }, 'action entering middleware');
    next(action);
};


export default
class Battle {
    store : any;

    constructor ( options : {} = {} ) {
        const MW_saga = createSagaMiddleware();

        this.store = createStore( 
            reducer,
            options.state || {},
            applyMiddleware( MW_logger, MW_saga )
        );
        MW_saga.run( sagas );
    }

    // init_game = payload => this.store.dispatch(
    //     Action.init_game(payload)
    // ) 

    get state () :BattleState { return this.store.getState() }

    dispatch( action : BattleAction ) :any {
        return this.store.dispatch(action);
    }

    dispatch_action( type: string, payload: any ) {
        return this.dispatch({ type, payload });
    }

    // init a game
    init( init_options : InitOptions ) {
        this.dispatch( actions.init_game(init_options) );
    }

};

/**
 * build middlewares for only certain types of actions
 * @param actions list of action types the middleware should be invoked for
 */
/*
function for_actions(...actions) {
    return store => next => action => {
        if (  actions.indexOf( action.type ) > -1 ) {
            this(store)(next)(action);
        }
        else {
            next(action);
        }
    };
}

const MW_firecons_fire = ( store => next => action => {
    let state = store.getState();

    next(action);

    _( _(state.objects).find({ id: action.object_id }) )
        .get( 'firecons', [] )
        .filter( f => f.target_id )
        .forEach( f => {
            let a = Action.firecon_fire({ object_id: action.object_id, firecon_id: f.id });
            console.log( 'a', a );
            store.dispatch(a);
        });
} ) ::for_actions( Action.FIRECONS_FIRE );

const MW_firecon_fire = ( store => next => action => {
    let state = store.getState();

    next(action);

    let ship = _.find( state.objects, { id: action.object_id } );

    let firecon = _.find( ship.firecons, { id: action.firecon_id } );

     _.get( ship, 'weapons', [] )
        .filter( f => f.firecon_id === firecon.id )
        .forEach( f => {
            let a = Action.weapon_fire({ object_id: action.object_id, weapon_id: f.id, target_id: firecon.target_id });
            store.dispatch(a);
        });
} ) ::for_actions( Action.FIRECON_FIRE );

function percolate_action() {
    return  store => next => action => {
        next(action);
        this(store)(next)(action);
    };
}

const find_object  = function(id) { return _.find( this.objects,  { id } ) };
const find_firecon = function(id) { return _.find( this.firecons, { id } ) };
const find_weapon  = function(id) { return _.find( this.weapons,  { id } ) };


let middlewares = [
    MW_firecons_fire,
    MW_firecon_fire,
];

import createSagaMiddleware from 'redux-saga';
import battle_saga from './saga';



*/
