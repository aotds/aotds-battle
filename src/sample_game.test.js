import {toBeDeepCloseTo, toMatchCloseTo } from 'jest-matcher-deep-close-to';
import fp from 'lodash/fp';

import { get_bogey } from './middlewares/selectors';

expect.extend({toBeDeepCloseTo, toMatchCloseTo });

import _ from 'lodash';
const debug = require('debug')('aotds:battle:test');
debug.inspectOpts.depth = 99;

import { cheatmode, rig_dice } from './dice';

import Battle from './index';

Date.prototype.toISOString = jest.fn( () => '2018-01-01' );

let turns = [ () => Promise.resolve( new Battle() ) ];

function wait_forever () {
    return new Promise((a,r) => {     });
}

turns[1] = (battle) => {

    const initial_state = require('./sample_game/initial_state').default;

    battle.dispatch_action( 'init_game', initial_state );

    expect(battle.state).toMatchObject({ 
        game: { name: 'gemini', turn: 0 },
        bogeys:{ enkidu: { name: 'Enkidu' },
            siduri: { name: 'Siduri' }, },
    });

    let enkidu_orders = { thrust: 1, turn: 1, bank: 1 };

    battle.dispatch_action( 'set_orders', 'enkidu', {
        navigation: enkidu_orders,
    } );

    battle.dispatch_action( 'play_turn' );
    
    // not yet...
    expect(battle.state).toHaveProperty( 'game.turn', 0 );

    battle.dispatch_action( 'set_orders', 'siduri', {
        navigation: { thrust: 1 },
    } );

    expect( battle.state.bogeys.enkidu.orders.navigation )
        .toMatchObject( enkidu_orders );

    // let's check the log
    expect( battle.state.log.map( l => l.type ).filter( t => !/@@/.test(t) ) ).toEqual([
        'INIT_GAME', 'SET_ORDERS', 'SET_ORDERS'
    ]);

    battle.dispatch_action('play_turn');

    expect(battle.state.game).toMatchObject({ turn: 1 });

    expect(battle.state).toMatchSnapshot();

    // orders cleared out 
    let still_with_orders = battle.state |> fp.get('bogeys')
        |> fp.values
        |> fp.filter('orders');

    expect(still_with_orders).toEqual([]);

    // Enkidu still have a drive section
    expect(battle.state.bogeys.enkidu).toHaveProperty(
        'drive.current'
    );

    const expect_close = val => val.map( v => x => expect(x).toBeCloseTo(v) );

    // ships have moved 
    expect( battle.state |> get_bogey('enkidu') |> fp.get('navigation') )
        .toMatchCloseTo({
            heading: 1,
            velocity: 1,
            coords: [ 1.5, 0.9  ],
        },1);

    expect( battle.state |> get_bogey('siduri') |> fp.get('navigation') )
        .toMatchCloseTo({
            heading: 6,
            velocity: 1,
            coords: [ 10, 9  ],
        },1);

        return battle;
};

turns[2] = function turn2(battle) {
    battle.dispatch_action( 'set_orders', 'enkidu', {
        firecons: { 1: { target_id: 'siduri' } }, 
        weapons: { 1: { firecon_id: 1 },
            2: { firecon_id: 1 }, 
            3: { firecon_id: 1 }, },
    });

    rig_dice([ 6, 5, 3, 3, 90, 90]);
    battle.dispatch_action( 'play_turn', true );

    expect( battle.state.bogeys.enkidu.weaponry.firecons[1] )
        .toMatchObject(
            { id: 1,  target_id: 'siduri' }
        );

    const enkidu = () => battle.state |> get_bogey('enkidu');
    const siduri = () => battle.state |> get_bogey('siduri');

    expect( enkidu() ).toHaveProperty( 'weaponry.weapons.1.firecon_id', 1 );

    let log = battle.state.log;
    log = log.splice( _.findLastIndex( log, { type: 'PLAY_TURN' } ) );

    expect( _.filter( log, { type: 'DAMAGE', bogey_id: 'siduri', penetrating: true, dice: [3] } ) ).toHaveLength(1);

    expect( siduri().structure.shields ).toMatchObject({
        1: { level: 1},
        2: { level: 2 }
    });

    expect( _.filter(log, { type: 'INTERNAL_DAMAGE' } ).length ).toBeGreaterThan(0)

    expect( siduri() )
        .toMatchObject({
            structure: {
                hull: { current: 3, max: 4 },
                armor: { current: 3, max: 4 },
            },
            drive: {
                current: 3,
                damage_level: 1,
                rating: 6,
            },
        });

    // only siduri gets damage
    expect( enkidu().structure )
        .toMatchObject({
            hull: { current: 4},
            armor: { current: 4},
        });

    return battle;
};

turns[3] = function turn3(battle) {
    // turn 3, we stop and fire like mad

    rig_dice([4,1]);

    [ 'enkidu', 'siduri' ].forEach( ship =>
        battle.dispatch_action( 'set_orders', ship, { navigation: { thrust: -1 }, } ) 
    );

    battle.play_turn(true);

    [ 'enkidu', 'siduri' ].forEach( ship => expect(
            get_object_by_id(battle.state,ship).navigation.velocity
        ).toEqual(0)
    );

    expect( get_object_by_id(battle.state,'siduri').drive )
        .toMatchObject({
            damage_level: 1,
            current: 3,
            thrust_used: 1,
        });

    return battle;
};

turns[4] = function turn4(battle) {

    // oh my, internal damages on the drive!
    rig_dice([6,1,6,1,5,90,90]);

    battle.play_turn(true);

    expect( get_object_by_id(battle.state,'siduri').structure )
        .toMatchObject({
            hull:  { current: 1, max: 4 },
            armor: { current: 2, max: 4 },
        });

    expect( get_object_by_id(battle.state,'siduri').drive )
        .toMatchObject({
            damage_level: 2,
            current: 0,
        });

    return battle;
};

turns[5] = function turn5(battle) {

    rig_dice([5,3]);

    battle.play_turn(true);

    expect( get_object_by_id(battle.state,'siduri').structure )
        .toMatchObject({
            hull:  { current: 1, max: 4 },
            armor: { current: 1, max: 4 },
        });

    return battle;
};

turns[6] = function turn6(battle) {

    rig_dice([4,6,2]);

    battle.play_turn(true);

    expect( get_object_by_id(battle.state,'siduri').structure )
        .toMatchObject({
            hull:  { current: 1, max: 4 },
            armor: { current: 0, max: 4 },
            status: "nominal"
        });

    return battle;
}

turns[7] = function turn7(battle) {

    rig_dice([5,2,90,90,90]);

    battle.play_turn(true);

    expect( get_object_by_id(battle.state,'siduri').structure )
        .toMatchObject({
            hull:  { current: 0, max: 4 },
            armor: { current: 0, max: 4 },
            status: "destroyed",
        });

    return battle;
}

turns[8] = function turn8(battle) {

    battle.play_turn(true);

    // siduri is gone
    expect( get_object_by_id(battle.state,'siduri') ).toBeUndefined();
    expect( get_object_by_id(battle.state,'enkidu') ).toBeDefined();

    expect( battle.state.log.map( l => l.type ) ).not.toContain( 'FIRE_WEAPON' );

    return battle;
}

turns.splice(4,99);

// turns.push( 
//     async function(battle) { return await wait_forever() }
// )

// jest.setTimeout(30000000);

let  previous = Promise.resolve();
turns = turns.map( t => {
    let p = previous.then(t);
    previous = p;
    return () => p;
});

turns.forEach( (t,i) => {
    test( `turn ${i}`, t );
});

