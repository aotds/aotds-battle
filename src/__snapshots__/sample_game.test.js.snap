// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`turn 0 1`] = `
Object {
  "bogeys": Object {
    "enkidu": Object {
      "drive": Object {
        "current": 6,
        "rating": 6,
      },
      "id": "enkidu",
      "name": "Enkidu",
      "navigation": Object {
        "coords": Array [
          0,
          0,
        ],
        "heading": 0,
        "velocity": 0,
      },
      "player_id": "yanick",
      "structure": Object {
        "armor": Object {
          "current": 4,
          "max": 4,
        },
        "hull": Object {
          "current": 4,
          "max": 4,
        },
        "shields": Object {
          "1": Object {
            "id": 1,
          },
          "2": Object {
            "id": 2,
          },
        },
        "status": "nominal",
      },
      "weaponry": Object {
        "firecons": Object {
          "1": Object {
            "id": 1,
          },
        },
        "weapons": Array [
          Object {
            "arcs": Array [
              "F",
            ],
            "class": 2,
            "type": "beam",
          },
          Object {
            "arcs": Array [
              "FS",
            ],
            "class": 1,
            "type": "beam",
          },
          Object {
            "arcs": Array [
              "FP",
            ],
            "class": 1,
            "type": "beam",
          },
        ],
      },
    },
    "siduri": Object {
      "drive": Object {
        "current": 6,
        "rating": 6,
      },
      "id": "siduri",
      "name": "Siduri",
      "navigation": Object {
        "coords": Array [
          10,
          10,
        ],
        "heading": 6,
        "velocity": 0,
      },
      "player_id": "yenzie",
      "structure": Object {
        "armor": Object {
          "current": 4,
          "max": 4,
        },
        "hull": Object {
          "current": 4,
          "max": 4,
        },
        "shields": Object {
          "1": Object {
            "id": 1,
          },
          "2": Object {
            "id": 2,
          },
        },
        "status": "nominal",
      },
      "weaponry": Object {},
    },
  },
  "game": Object {
    "action_stack": Array [],
    "name": "gemini",
    "next_action_id": 14,
    "players": Array [
      Object {
        "id": "yanick",
      },
      Object {
        "id": "yenzie",
      },
    ],
    "turn": 1,
    "turn_times": Object {
      "started": "2018-01-01",
    },
  },
  "log": Array [
    Object {
      "bogeys": Array [
        Object {
          "drive": Object {
            "current": 6,
            "rating": 6,
          },
          "id": "enkidu",
          "name": "Enkidu",
          "navigation": Object {
            "coords": Array [
              0,
              0,
            ],
            "heading": 0,
            "velocity": 0,
          },
          "player_id": "yanick",
          "structure": Object {
            "armor": 4,
            "hull": 4,
            "shields": Array [
              1,
              2,
            ],
            "status": "nominal",
          },
          "weaponry": Object {
            "firecons": 1,
            "weapons": Array [
              Object {
                "arcs": Array [
                  "F",
                ],
                "class": 2,
                "type": "beam",
              },
              Object {
                "arcs": Array [
                  "FS",
                ],
                "class": 1,
                "type": "beam",
              },
              Object {
                "arcs": Array [
                  "FP",
                ],
                "class": 1,
                "type": "beam",
              },
            ],
          },
        },
        Object {
          "drive": Object {
            "current": 6,
            "rating": 6,
          },
          "id": "siduri",
          "name": "Siduri",
          "navigation": Object {
            "coords": Array [
              10,
              10,
            ],
            "heading": 6,
            "velocity": 0,
          },
          "player_id": "yenzie",
          "structure": Object {
            "armor": 4,
            "hull": 4,
            "shields": Array [
              1,
              2,
            ],
            "status": "nominal",
          },
        },
      ],
      "game": Object {
        "name": "gemini",
        "players": Array [
          Object {
            "id": "yanick",
          },
          Object {
            "id": "yenzie",
          },
        ],
      },
      "meta": Object {
        "id": 1,
        "timestamp": "2018-01-01",
      },
      "type": "INIT_GAME",
    },
    Object {
      "bogey_id": "enkidu",
      "meta": Object {
        "id": 2,
        "timestamp": "2018-01-01",
      },
      "orders": Object {
        "navigation": Object {
          "bank": 1,
          "thrust": 1,
          "turn": 1,
        },
      },
      "type": "SET_ORDERS",
    },
    Object {
      "bogey_id": "siduri",
      "meta": Object {
        "id": 4,
        "timestamp": "2018-01-01",
      },
      "orders": Object {
        "navigation": Object {
          "thrust": 1,
        },
      },
      "type": "SET_ORDERS",
    },
    Object {
      "force": false,
      "meta": Object {
        "child_actions": Array [
          Object {
            "meta": Object {
              "id": 7,
              "parent_action_id": 5,
              "timestamp": "2018-01-01",
            },
            "type": "MOVEMENT_PHASE",
          },
          Object {
            "meta": Object {
              "id": 8,
              "parent_action_id": 5,
              "timestamp": "2018-01-01",
            },
            "type": "EXECUTE_FIRECON_ORDERS",
          },
          Object {
            "meta": Object {
              "id": 9,
              "parent_action_id": 5,
              "timestamp": "2018-01-01",
            },
            "type": "ASSIGN_WEAPONS_TO_FIRECONS",
          },
          Object {
            "meta": Object {
              "id": 10,
              "parent_action_id": 5,
              "timestamp": "2018-01-01",
            },
            "type": "EXECUTE_FIRECON_ORDERS",
          },
          Object {
            "meta": Object {
              "id": 11,
              "parent_action_id": 5,
              "timestamp": "2018-01-01",
            },
            "type": "FIRE_WEAPONS",
          },
          Object {
            "meta": Object {
              "id": 12,
              "parent_action_id": 5,
              "timestamp": "2018-01-01",
            },
            "type": "CLEAR_ORDERS",
          },
        ],
        "id": 5,
        "timestamp": "2018-01-01",
      },
      "type": "PLAY_TURN",
    },
  ],
}
`;
